modules/internal/DefaultRectangular.chpl

```diff
@@ -842,24 +841,31 @@ module DefaultRectangular {
           sum += (ind(i) - off(i)) * blk(i) / abs(str(i)):idxType;
         return sum;
       } else {
-        var sum = if earlyShiftData then 0:idxType else origin;
-
-        // If the user asserts that there is no slicing in their program,
-        // then blk(rank) == 1. Knowing this, we need not multiply the final
-        // ind(...) by anything. This may lead to performance improvements for
-        // array accesses.
-        if assertNoSlicing {
-          for param i in 1..rank-1 {
-            sum += ind(i) * blk(i);
+        if (rank == 1 && earlyShiftData) {
+          if __primitive("optimize_array_blk_mult") {
+            return ind(1);
+          } else {
+            return ind(1) * blk(1);
```

Comment:
I didn't see any performance impact from it, but it has a significant impact on the generated code readability

```chapel
for i in 1..10 do A[i] = B[i] + C[i];
```

```c
  // w/o rank == 1
  for (i = INT64(1); ((i <= _ic__F1_high)); i += INT64(1)) {
    sum = INT64(0);
    sum += i;
    call_tmp7 = (coerce_tmp + sum);
    sum2 = INT64(0);
    sum2 += i;
    call_tmp9 = (coerce_tmp2 + sum2);
    ret = *(call_tmp9);
    sum3 = INT64(0);
    sum3 += i;
    call_tmp11 = (coerce_tmp3 + sum3);
    *(call_tmp7) = ((int64_t)((ret + *(call_tmp11))));
  }
```
vs 
```c
  //w/ rank == 1
  for (i = INT64(1); ((i <= _ic__F1_high)); i += INT64(1)) {
    call_tmp7 = (coerce_tmp + i);
    call_tmp9 = (coerce_tmp2 + i);
    call_tmp11 = (coerce_tmp3 + i);
    *(call_tmp7) = ((int64_t)((*(call_tmp9) + *(call_tmp11))));
  }
```

The backend should have no problem optimizing what we would generate without the rank==1 special case, but I think the generated code improvement is worthwhile.